yash-env/src/trap.rs

```diff
@@ -222,24 +250,31 @@
 
         for (&cond, state) in &mut self.traps {
             let option = match cond {
-                Condition::Signal(Signal::SIGCHLD) => EnterSubshellOption::KeepInternalHandler,
-                Condition::Signal(Signal::SIGINT | Signal::SIGQUIT) if ignore_sigint_sigquit => {
-                    EnterSubshellOption::Ignore
+                Condition::Exit => EnterSubshellOption::ClearInternalHandler,
+                Condition::Signal(number) => {
+                    use signal::Name::*;
+                    match system.signal_name_from_number(number) {
+                        Chld => EnterSubshellOption::KeepInternalHandler,
+                        Int | Quit if ignore_sigint_sigquit => EnterSubshellOption::Ignore,
+                        Tstp | Ttin | Ttou
+                            if keep_stopper_handlers
+                                && state.internal_handler() != SignalHandling::Default =>
+                        {
+                            EnterSubshellOption::Ignore
+                        }
+                        _ => EnterSubshellOption::ClearInternalHandler,
+                    }
```

Comment:
Sorry, your suggestion looks more complicated than before. The conditions `ignore_sigint_sigquit` and `keep_stopper_handlers` are relevant with a specific match branch, so it would be more intuitive when they are tied with the branch rather than the whole match expression.